board_and_pieces.py

```diff
@@ -1,53 +1,126 @@
+        x = self.x
+        y = self.y
+        all_squares = [(x+1, y-2), (x+2, y-1), (x+2, y+1), (x+1, y+2),
+                       (x-1, y-2), (x-2, y-1), (x-1, y+2), (x-2, y+1)]
+        # filter out moves that are not even on the board
+        moves_on_board = [
+            square for square in all_squares if self.on_board(square)]
+
+        valid_moves = [move for move in moves_on_board if (
+            not move in board.board_dict) or (board.board_dict[move].colour != self.colour)]
+
+        return valid_moves
 
 
 class Bishop(Piece):
-    def __init__(self, x: int, y: int, colour: Enum):
+    def __init__(self, x: int, y: int, colour: Enum, has_moved=False):
         super().__init__(x, y, colour)
-        self.letter = "B" if colour == Colour.WHITE else "b"
+        self.letter = "♗" if colour == Colour.BLACK else "♝"
+
+    def get_valid_moves(self, board, last_moved: str, initial_pos: str, final_pos: str):
+        return self.increment([(1, -1), (1, 1), (-1, 1), (-1, -1)], board)
 
 
 class Rook(Piece):
-    def __init__(self, x: int, y: int, colour: Enum):
+    def __init__(self, x: int, y: int, colour: Enum, has_moved=False):
         super().__init__(x, y, colour)
-        self.letter = "R" if colour == Colour.WHITE else "r"
+        self.letter = "♖" if colour == Colour.BLACK else "♜"
+
+    def get_valid_moves(self, board, last_moved: str, initial_pos: str, final_pos: str):
+        return self.increment([(1, 0), (0, 1), (-1, 0), (0, -1)], board)
 
 
 class Queen(Piece):
-    def __init__(self, x: int, y: int, colour: Enum):
+    def __init__(self, x: int, y: int, colour: Enum, has_moved=False):
         super().__init__(x, y, colour)
-        self.letter = "Q" if colour == Colour.WHITE else "q"
+        self.letter = "♕" if colour == Colour.BLACK else "♛"
+
+    def get_valid_moves(self, board, last_moved: str, initial_pos: str, final_pos: str):
+        return self.increment([(1, 0), (0, 1), (-1, 0), (0, -1), (1, -1), (1, 1), (-1, 1), (-1, -1)], board)
 
 
 class King(Piece):
-    def __init__(self, x: int, y: int, colour: Enum):
+    def __init__(self, x: int, y: int, colour: Enum, has_moved=False):
         super().__init__(x, y, colour)
-        self.letter = "K" if colour == Colour.WHITE else "k"
+        self.letter = "♔" if colour == Colour.BLACK else "♚"
+
+    def get_valid_moves(self, board, last_moved: str, initial_pos: str, final_pos: str):
+        x = self.x
+        y = self.y
 
 
 class Pawn(Piece):
-    def __init__(self, x: int, y: int, colour: Enum):
+    def __init__(self, x: int, y: int, colour: Enum, has_moved=False):
         super().__init__(x, y, colour)
-        self.letter = "P" if colour == Colour.WHITE else "p"
+        self.letter = "♙" if colour == Colour.BLACK else "♟︎"
+        self.direction_factor = 1 if colour == Colour.BLACK else -1
+
+    def get_valid_moves(self, board, last_moved: str, initial_pos: str, final_pos: str):
```

Comment:
should initial_pos, final_pos be array types?

Reply:
Oh rip they're tuples, don't know why they're put as str